src/Materials/materialHelper.ts

```diff
@@ -577,7 +577,8 @@ export class MaterialHelper {
      * @param defines The current Defines of the effect
      */
     public static PrepareAttributesForInstances(attribs: string[], defines: any): void {
-        if (defines["INSTANCES"]) {
+        // Check if we are handling a MaterialDefine or an Array containing the #define
+        if (defines["INSTANCES"] || Array.isArray(defines)) {

```

Comment:
Not sure to get why this test :)
what if I call PrepareAttributesForInstances with a defines not containing INSTANCES?

Reply:
Haha, maybe I need to catch up on some sleep, let me try to formulate what I think is going on and see if it makes sense. :D

We have two cases of defines that could arrive in this function as of now:
One is defines as MaterialDefines which gets handled by the defines["INSTANCES"] here. 
(Coming from most Materials like shadowOnlyMaterial)
Those classes all call PrepareAttributesForInstances without checking if it should be called, so this check we can't live without.

The other source of defines is coming in as an array filled with strings.
(Coming from the newly changed cases like EffectsLayer)
Those classes all check "if (useInstances)" before calling the helper function, since world0-3 were inline, to begin with.
Initially, I wanted to check if the correct array define is present using indexOf, but due to the performance overhead, we removed it.

However, if we only check defines["INSTANCES"], it will return undefined for any of those string define arrays, so we would never get to apply the attributes in that case. 
Without indexof or a faster alternative to look inside the array we now have to rely on MaterialDefines getting checked properly and giving standard arrays a free pass so to speak, but since those are pre-checked we are currently sure it needs to be applied.

Admittedly not too pretty, that's why I was a bit confused about removing it and relying only on the array type check in my last post :)

The only other option I see as of now to keep it all in one place would be an optional parameter like forceApply, so we could call. PrepareAttributesForInstances(attribs, null, true). Not sure if this is much better though. 

I guess if we want unified attribute management globally we would need to change those classes to use real MaterialDefines anyway so we might want to avoid touching the function parameters.

In case this is too hacky let's scrap the PR and move back to the drawing board to figure out a better way or a bigger milestone that encapsulates more than just this smallish change. I meant for it to be the first change in a series of things to adjust to make the bigger picture happen anyway, like we talked about on the forums.

I'd really love to stick with this topic and get some nice features in (like the instanciated mesh morphs) if possible, but it doesn't have to be today or tomorrow :) If we can start off with this that's cool, if not, no hard feelings, I learned a lot :)

Ugh, sorry for writing a book...

Reply:
No worries ;) I now get why so this is fine!

In this case, we should just have two functions. One with the check that will call the second one where we apply the values.

The 2nd one will then be called by those who do not need a check and performance will be happy

Reply:
Feels way better 👍 Thx!